diff --git a/Makefile b/Makefile
index 7b60eb103c5d..02edc63279ed 100644
--- a/Makefile
+++ b/Makefile
@@ -743,6 +743,12 @@ CFLAGS_GCOV	+= -fno-tree-loop-im
 endif
 export CFLAGS_GCOV
 
+ifdef CONFIG_NO_AUTO_INLINE
+KBUILD_CFLAGS   += $(call cc-option, -fno-inline-functions) \
+          $(call cc-option, -fno-inline-small-functions) \
+          $(call cc-option, -fno-inline-functions-called-once)
+endif
+
 # The arch Makefiles can override CC_FLAGS_FTRACE. We may also append it later.
 ifdef CONFIG_FUNCTION_TRACER
   CC_FLAGS_FTRACE := -pg
@@ -799,13 +805,22 @@ endif # need-config
 
 KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 
+ifdef CONFIG_CC_OPTIMIZE_FOR_DEBUGGING
+KBUILD_CFLAGS  += $(call cc-option, -Og)
+KBUILD_CFLAGS  += $(call cc-option, -Wno-unused-variable)
+KBUILD_CFLAGS   += $(call cc-disable-warning,maybe-uninitialized,)
+KBUILD_CFLAGS   += $(call cc-disable-warning,unused-function,)
+else
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
 KBUILD_RUSTFLAGS += -Copt-level=2
-else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
+else
+ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
 KBUILD_RUSTFLAGS += -Copt-level=s
 endif
+endif
+endif
 
 # Always set `debug-assertions` and `overflow-checks` because their default
 # depends on `opt-level` and `debug-assertions`, respectively.
diff --git a/arch/arm/mm/mmu.c b/arch/arm/mm/mmu.c
index 3f774856ca67..eacceafa2044 100644
--- a/arch/arm/mm/mmu.c
+++ b/arch/arm/mm/mmu.c
@@ -1738,7 +1738,7 @@ static void __init early_fixmap_shutdown(void)
 		pte_t *pte;
 		struct map_desc map;
 
-		map.virtual = fix_to_virt(i);
+		map.virtual = __fix_to_virt(i);
 		pte = pte_offset_early_fixmap(pmd_off_k(map.virtual), map.virtual);
 
 		/* Only i/o device mappings are supported ATM */
diff --git a/include/linux/bpf.h b/include/linux/bpf.h
index 3b94ec161e8c..2e6d99ef5d0d 100644
--- a/include/linux/bpf.h
+++ b/include/linux/bpf.h
@@ -1231,11 +1231,7 @@ struct bpf_dispatcher {
 #endif
 };
 
-#ifndef __bpfcall
-#define __bpfcall __nocfi
-#endif
-
-static __always_inline __bpfcall unsigned int bpf_dispatcher_nop_func(
+static __nocfi unsigned int bpf_dispatcher_nop_func(
 	const void *ctx,
 	const struct bpf_insn *insnsi,
 	bpf_func_t bpf_func)
diff --git a/include/linux/compiler_attributes.h b/include/linux/compiler_attributes.h
index 32284cd26d52..54969838ae5f 100644
--- a/include/linux/compiler_attributes.h
+++ b/include/linux/compiler_attributes.h
@@ -147,7 +147,7 @@
  *
  *   gcc: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-error-function-attribute
  */
-#if __has_attribute(__error__)
+#if __has_attribute(__error__) && !defined(CONFIG_CC_OPTIMIZE_FOR_DEBUGGING)
 # define __compiletime_error(msg)       __attribute__((__error__(msg)))
 #else
 # define __compiletime_error(msg)
@@ -385,7 +385,8 @@
  *
  *   gcc: https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-warning-function-attribute
  */
-#if __has_attribute(__warning__)
+//#if __has_attribute(__warning__)
+#if __has_attribute(__warning__) && !defined(CONFIG_CC_OPTIMIZE_FOR_DEBUGGING)
 # define __compiletime_warning(msg)     __attribute__((__warning__(msg)))
 #else
 # define __compiletime_warning(msg)
diff --git a/include/linux/compiler_types.h b/include/linux/compiler_types.h
index f14c275950b5..430e04c71216 100644
--- a/include/linux/compiler_types.h
+++ b/include/linux/compiler_types.h
@@ -477,6 +477,7 @@ struct ftrace_likely_data {
 	(sizeof(t) == sizeof(char) || sizeof(t) == sizeof(short) || \
 	 sizeof(t) == sizeof(int) || sizeof(t) == sizeof(long))
 
+#if !defined(CONFIG_CC_OPTIMIZE_FOR_DEBUGGING)
 #ifdef __OPTIMIZE__
 # define __compiletime_assert(condition, msg, prefix, suffix)		\
 	do {								\
@@ -493,6 +494,9 @@ struct ftrace_likely_data {
 #else
 # define __compiletime_assert(condition, msg, prefix, suffix) do { } while (0)
 #endif
+# else
+# define __compiletime_assert(condition, msg, prefix, suffix) do { } while (0)
+#endif
 
 #define _compiletime_assert(condition, msg, prefix, suffix) \
 	__compiletime_assert(condition, msg, prefix, suffix)
diff --git a/init/Kconfig b/init/Kconfig
index 5783a0b87517..ad395d0f569f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1400,6 +1400,25 @@ choice
 	prompt "Compiler optimization level"
 	default CC_OPTIMIZE_FOR_PERFORMANCE
 
+config CC_OPTIMIZE_FOR_DEBUGGING
+	bool "Optimize for better debugging experience (-Og)"
+	select NO_AUTO_INLINE
+	help
+	  This will apply GCC '-Og' optimization level which is supported
+	  since GCC 4.8. This optimization level offers a reasonable level
+	  of optimization while maintaining fast compilation and a good
+	  debugging experience. It is similar to '-O1' while preferring to
+	  keep debug ability over runtime speed. The overall performance
+	  will drop a bit (~6%).
+
+	  Use only if you want to debug the kernel, especially if you want
+	  to have better kernel debugging experience with gdb facilities
+	  like kgdb or qemu. If enabling this option breaks your kernel,
+	  you should either disable this or find a fix (mostly in the arch
+	  code).
+
+	  If unsure, select N.
+
 config CC_OPTIMIZE_FOR_PERFORMANCE
 	bool "Optimize for performance (-O2)"
 	help
@@ -1455,6 +1474,7 @@ config LD_ORPHAN_WARN_LEVEL
         default "error" if WERROR
         default "warn"
 
+
 config SYSCTL
 	bool
 
diff --git a/kernel/module/tree_lookup.c b/kernel/module/tree_lookup.c
index 277197977d43..02d24de0f661 100644
--- a/kernel/module/tree_lookup.c
+++ b/kernel/module/tree_lookup.c
@@ -33,13 +33,13 @@ static __always_inline unsigned long __mod_tree_size(struct latch_tree_node *n)
 	return (unsigned long)mod_mem->size;
 }
 
-static __always_inline bool
+static bool
 mod_tree_less(struct latch_tree_node *a, struct latch_tree_node *b)
 {
 	return __mod_tree_val(a) < __mod_tree_val(b);
 }
 
-static __always_inline int
+static int
 mod_tree_comp(void *key, struct latch_tree_node *n)
 {
 	unsigned long val = (unsigned long)key;
diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index a30c03a66172..fd1c93937c61 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -433,6 +433,30 @@ config GDB_SCRIPTS
 	  additional functions are available to analyze a Linux kernel
 	  instance. See Documentation/dev-tools/gdb-kernel-debugging.rst
 	  for further details.
+config NO_AUTO_INLINE
+	bool "Disable compiler auto-inline optimizations"
+	help
+	  This will prevent the compiler from optimizing the kernel by
+	  auto-inlining functions not marked with the inline keyword.
+	  With this option, only functions explicitly marked with
+	  "inline" will be inlined. This will allow the function tracer
+	  to trace more functions because it only traces functions that
+	  the compiler has not inlined.
+
+	  Enabling this function can help debugging a kernel if using
+	  the function tracer. But it can also change how the kernel
+	  works, because inlining functions may change the timing,
+	  which could make it difficult while debugging race conditions.
+
+	  If unsure, select N.
+
+config ENABLE_WARN_DEPRECATED
+	bool "Enable __deprecated logic"
+	default y
+	help
+	  Enable the __deprecated logic in the kernel build.
+	  Disable this to suppress the "warning: 'foo' is deprecated
+	  (declared at kernel/power/somefile.c:1234)" messages.
 
 endif # DEBUG_INFO
 
